MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c

```diff
@@ -1033,9 +1089,41 @@ NvmExpressDriverBindingStart (
 
     // MU_CHANGE - Support alternative hardware queue sizes in NVME driver
 
+    // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+    //
+    // Set the Admin Queue Atttributes
+    //
+    Aqa = AllocateZeroPool (sizeof (NVME_AQA));
+
+    if (Aqa == NULL) {
+      DEBUG ((DEBUG_ERROR, "NvmExpressDriverBindingStart: allocating pool for Nvme Aqa Data failed!\n"));
+      Status = EFI_OUT_OF_RESOURCES;
+      goto Exit;
+    }
+
+    // Set the sizes of the admin submission & completion queues in number of entries
+    Aqa->Asqs  = PcdGetBool (PcdSupportAlternativeQueueSize) ? MIN (NVME_ALTERNATIVE_MAX_QUEUE_SIZE, Private->Cap.Mqes) : NVME_ASQ_SIZE;
+    Aqa->Rsvd1 = 0;
+    Aqa->Acqs  = PcdGetBool (PcdSupportAlternativeQueueSize) ? MIN (NVME_ALTERNATIVE_MAX_QUEUE_SIZE, Private->Cap.Mqes) : NVME_ACQ_SIZE;
+    Aqa->Rsvd2 = 0;
+
+    //
+    // Save Queue Pair Data for admin queues in controller data structure
+    //
+    Private->SqData[0].NumberOfEntries = Aqa->Asqs;
+    Private->CqData[0].NumberOfEntries = Aqa->Acqs;
+
+    //
+    // Set admin queue entry size to default
     //
-    // Depending on PCD disablement, either support the default or alternative
-    // queue sizes.
+    Private->SqData[0].EntrySize = NVME_IOSQES_MIN;
+    Private->CqData[0].EntrySize = NVME_IOCQES_MIN;
+
+    // Calculate the number of pages required for the admin queues
+    // TODO create helper functions for calculating num pages for a queue.
+    AdminQueuePageCount = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * LShiftU64 (2, Private->SqData[0].EntrySize))
+                          + EFI_SIZE_TO_PAGES (Private->CqData[0].NumberOfEntries * LShiftU64 (2, CqData[0].EntrySize));

```

Comment:
```suggestion
    AdminQueuePageCount = EFI_SIZE_TO_PAGES ((Private->SqData[0].NumberOfEntries * LShiftU64 (2, Private->SqData[0].EntrySize))
                          + (Private->CqData[0].NumberOfEntries * LShiftU64 (2, CqData[0].EntrySize));

```
This would potentially waste less memory since each calculation is rounding to the nearest page, but if the sizes together could fit within the same page, we would allocate an extra page.

Reply:
In `NvmeControllerInitAdminQueues` we have to write the base address of both the submission and completion queues.  Are you suggesting the base address of the completion queue starts in the middle of a page after the end of the submission queue? This deviates from previous behavior where each queue was allocated space in units of whole pages. In NvmExpress.c, this is described in a comment where for example, the 1st 4kB boundary is the start of the admin submission queue, and the 5th 4kB boundary is the start of the completion queue. 

Btw maybe the variable would be better named as "AdminQueuePairPageCount"

Reply:
Gotcha, I missed both queues are pulling from the same buffer.